examples/dfdl-bigtable-pubsub-example/src/main/java/com/example/dfdl/PubSubServer.java

```diff
@@ -0,0 +1,150 @@
+ * the message in a Spring Channel - The ServiceActivator processes the received messages in a
+ * Spring Channel The sender: - The Message Gateway will write messages to the Spring Channel; - The
+ * ServiceActivator will consume the messages in the Spring Channel and sends them to the GCP Topic;
+ * - The Output Channel Adapter will ensure that the messages are delivered to the GCP Topic.
+ */
+@SpringBootApplication
+public class PubSubServer {
+
+  private static final Log LOGGER = LogFactory.getLog(PubSubServer.class);
+
+  @Value("${pubsub.data.binary.subscription}")
+  String pubsubDataBinarySub;
+
+  @Value("${pubsub.data.json.topic}")
+  String pubsubDataJsonTopic;
+
+  @Value("${dfdl.definition.name}")
+  String dfdlDefName;
+
+  @Autowired private DfdlService dfdlService;
+  @Autowired private BigtableService bigtableService;
+  @Autowired private PubsubOutboundGateway messagingGateway;
+
+  /** Returns a channel where the adapter sends the received messages. */
+  @Bean
+  public MessageChannel inputChannel() {
+    return new DirectChannel();
+  }
+
+  /**
+   * Returns an inbound channel adapter to listen to a subscription and send messages to the input
+   * message channel.
+   */
+  @Bean
+  public PubSubInboundChannelAdapter inboundChannelAdapter(
+      @Qualifier("inputChannel") MessageChannel messageChannel, PubSubTemplate pubSubTemplate) {
+    PubSubInboundChannelAdapter adapter =
+        new PubSubInboundChannelAdapter(pubSubTemplate, pubsubDataBinarySub);
+    adapter.setOutputChannel(messageChannel);
+    adapter.setAckMode(AckMode.MANUAL);
+    adapter.setPayloadType(String.class);
+    return adapter;
+  }
+
+  /**
+   * Processes the received message in a spring channel. The received message is the one to be
+   * interpreted or transform using the dfdl definition and be republished in a new topic.
+   */
+  @Bean
+  @ServiceActivator(inputChannel = "inputChannel")
+  public MessageHandler messageReceiver() {
+    return message -> {
+      LOGGER.info("Message arrived! Payload: " + message.getPayload());
+      DfdlDef dfdlDef;
+      try {
+        // Get DFDL Definition from Bigtable.
+        dfdlDef = bigtableService.getDfdlDef(dfdlDefName);
+        System.out.println("Definition from Bigtable: " + dfdlDef.getDefinition());
+
+        // Transform message using the dfdl definition.
+        String messageConverted =
+            dfdlService.convertDataMessage((String) message.getPayload(), dfdlDef);
+
+        // Republish the message in json format in a new topic
+        messagingGateway.sendToPubsub(pubsubDataJsonTopic, messageConverted);
+
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+      BasicAcknowledgeablePubsubMessage originalMessage =
+          message
+              .getHeaders()
+              .get(GcpPubSubHeaders.ORIGINAL_MESSAGE, BasicAcknowledgeablePubsubMessage.class);
+      originalMessage.ack();
+    };
```

Comment:
The ack of the original message will happen even if the lookup in Bigtable or the publish of the transformed message fail. Should not the ack happen only if the message can be successfully processed?